kernel-rs/src/fs/lfs/inode.rs

```diff
@@ -256,40 +256,87 @@ impl InodeGuard<'_, Lfs> {
+    ///
+    /// * If you do not need to write to the block, use `InodeGuard::bmap_read` instead.
+    /// * After writing to the `Buf`, you should commit the segment (if it is full),
+    /// and then call `InodeGuard::update`.
+    ///
+    /// # Inode content
     ///
     /// The content (data) associated with each inode is stored
     /// in blocks on the disk. The first NDIRECT block numbers
     /// are listed in self->addrs[].  The next NINDIRECT blocks are
     /// listed in block self->addr_indirect.
     /// Return the disk block address of the nth block in inode self.
     /// If there is no such block, bmap allocates one.
-    #[allow(dead_code)]
-    pub fn bmap_or_alloc(&mut self, bn: usize, tx: &Tx<'_, Lfs>, ctx: &KernelCtx<'_, '_>) -> u32 {
-        self.bmap_internal(bn, Some(tx), ctx)
+    pub fn write_data_block(
+        &mut self,
+        bn: usize,
+        tx: &Tx<'_, Lfs>,
+        ctx: &KernelCtx<'_, '_>,
+    ) -> Buf {
+        self.bmap_internal(bn, true, Some(tx), ctx)
     }
 
-    pub fn bmap(&mut self, bn: usize, ctx: &KernelCtx<'_, '_>) -> u32 {
-        self.bmap_internal(bn, None, ctx)
+    /// Returns a `Buf` that has the inode's `bn`th data block content.
+    ///
+    /// # Note
+    ///
+    /// Use the returned `Buf` only to read the inode's data block's content.
+    /// Any write operations to a inode's data block should be done using `InodeGuard::bmap_write`.
+    pub fn read_data_block(&mut self, bn: usize, ctx: &KernelCtx<'_, '_>) -> Buf {
+        self.bmap_internal(bn, false, None, ctx)
     }
 
     fn bmap_internal(
         &mut self,
         bn: usize,
+        write: bool,
         tx_opt: Option<&Tx<'_, Lfs>>,
         ctx: &KernelCtx<'_, '_>,
-    ) -> u32 {
-        let inner = self.deref_inner();
+    ) -> Buf {
+        let dev = self.dev;
+        let inum = self.inum;
+        let inner = self.deref_inner_mut();
 
         if bn < NDIRECT {
             let addr = inner.addr_direct[bn];
-            if addr == 0 {
-                let (_, addr) = tx_opt
-                    .expect("bmap: out of range")
-                    .balloc(self.inum, bn as u32, ctx);
-                self.deref_inner_mut().addr_direct[bn] = addr;
+            if write {
+                let tx = tx_opt.expect("bmap: out of range");
+                let (buf, addr) = if addr == 0 {
+                    // Allocate an empty block.
+                    tx.balloc(inum, bn as u32, ctx)
+                } else {
+                    // Copy from old block to new block.
+                    let (mut buf, new_addr) = tx
+                        .fs
+                        .segment()
+                        .get_or_add_data_block(inum, bn as u32, ctx)
+                        .unwrap();
+                    let old_buf = hal().disk().read(dev, addr, ctx);
+                    // SAFETY: The old data block's content will not be used from now on.
+                    unsafe {
+                        core::ptr::copy(
```

Comment:
Does this code mean that data inside `old_buf` is written on `buf`? Is it equivalent to write something on hal().disk()?

Reply:
Yes, the data on `old_buf` is just being copied to the `buf` here.
You can think of `ptr::copy` something similar to `memcpy` in C.